rgw: multisite enabled over multiple clusters

alimaredia opened the pull request:
Added '--cluster' to all necessary commands
ex: radosgw-admin, rados, ceph, made sure
necessary checks were in place so that clients
can be read with our without a cluster_name
preceeding them

Made master_client defined in the config for
radosgw-admin task

Signed-off-by: Ali Maredia <amaredia@redhat.com>

oritwas commented:
I don't like duplicate parameters and the configuration already has the master zone client.
you can use get_config_master_client  from task/utils/rgw.py

cbodley commented:
@alimaredia ping

oritwas commented on qa/suites/rgw/multisite/all/multi_cluster.yaml:
```diff
@@ -0,0 +1,84 @@
+  - cluster1.osd.1
+  - cluster1.osd.2
+  - cluster2.mon.c
+  - cluster1.client.0
```
it will be easier to read if the cluster weren't mixed also order by component

oritwas commented on qa/suites/rgw/multisite/all/multi_cluster.yaml:
```diff
@@ -0,0 +1,84 @@
+    conf:
+      client.0:
+        rgw region: jerry
+        rgw zone: seinfeld
```
great name :+1:

oritwas replied:
we will need to clean up the region at some point (not in this pr)

oritwas commented on qa/tasks/radosgw_admin.py:
```diff
@@ -63,28 +62,30 @@ def remove_newlines(string):
 
-    multi_region_run = rgw_utils.multi_region_enabled(ctx)
+    # regions and config found from rgw task
+    regions = ctx.rgw.regions
```
check to see there are regions

oritwas commented on qa/tasks/radosgw_admin.py:
```diff
@@ -63,28 +62,30 @@ def remove_newlines(string):
-    multi_region_run = rgw_utils.multi_region_enabled(ctx)
+    # regions and config found from rgw task
+    regions = ctx.rgw.regions
+    config = ctx.rgw.config
```
check to see there is a config

oritwas commented on qa/tasks/radosgw_agent.py:
```diff
@@ -18,8 +18,9 @@ def run_radosgw_agent(ctx, config):
         # don't process entries that are not clients
-        if not client.startswith('client.'):
-            log.debug('key {data} does not start with \'client.\', moving on'.format(
+    	log.debug("client is %r", client)
```
extra space

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -62,19 +39,22 @@ def create_apache_dirs(ctx, config, on_client = None, except_client = None):
             continue
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_cluster = cluster_name + '.' + daemon_type + '.' + client_id
```
What is the cluster name is empty?

alimaredia replied:
It defaults to "ceph"

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -134,6 +114,10 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_id = daemon_type + '.' + client_id
+    	client_with_cluster = cluster_name + '.' + client_with_id
```
what if cluster_name is empty?

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
     Start rgw on remote sites.
     """
-    log.info('Starting rgw...')
+    log.info('ALI ADDED Starting rgw...')
```
don't forget to remove this :)

alimaredia replied:
Yes i plan to do a cleanup of the entire PR. I just add that in so I can easily ctrl+f important parts of the logs.

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
+        cluster_name, daemon_type, client_id = teuthology.split_role(client)
+        client_with_id = daemon_type + '.' + client_id
+        # in case client doesn't have cluster_name attached to it
+        client_with_cluster = cluster_name + '.' + client_with_id
```
see previous comment

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -134,6 +114,10 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
             continue
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_id = daemon_type + '.' + client_id
```
I think the comment is wrong , you need if for when you provide the cluster separately

oritwas replied:
Also please check that daemon_type and client_id are not empty

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -177,32 +161,33 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
             fcgi_config = f.read()
         with file(src, 'rb') as f:
             conf = f.read() + fcgi_config
+            log.debug('ALI conf is %r', conf)
```
please remove

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -217,21 +202,21 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
             data="""#!/bin/sh
 ulimit -c unlimited
-exec radosgw -f -n {client} -k /etc/ceph/ceph.{client}.keyring {rgw_options}
+exec radosgw -f -n {client_with_id} --cluster {cluster_name} -k /etc/ceph/{client_with_cluster}.keyring {rgw_options}
```
will it be simpler to use --id instead of -n?

alimaredia replied:
im not sure, does it make a difference? @cbodley have an opinion?

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
+        client_with_cluster = cluster_name + '.' + client_with_id
         zone = rgw_utils.zone_for_client(ctx, client)
         log.debug('zone %s', zone)
+
```
nit empty line

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -393,6 +386,8 @@ def start_apache(ctx, config, on_client = None, except_client = None):
         clients_to_run = config.keys()
     for client in clients_to_run:
+        cluster_name, daemon_type, client_id = teuthology.split_role(client)
+        client_with_cluster = cluster_name + '.' + daemon_type + '.' + client_id
```
please check for empty cluster_name and daemon_type

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -636,6 +541,7 @@ def configure_users_for_client(ctx, config, client, everywhere=False):
             continue
 
         for client_name in clients_to_create_as:
+            cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
not used, please remove

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -676,6 +582,7 @@ def configure_users(ctx, config,  everywhere=False):
         if everywhere:
             clients_to_create_as = config.keys()
         for client_name in clients_to_create_as:
+            cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
not used, please remove

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -738,6 +647,7 @@ def configure_multisite_regions_and_zones(ctx, config, regions, role_endpoints,
     log.debug('regions are %r', regions)
     log.debug('role_endpoints = %r', role_endpoints)
     log.debug('realm is %r', realm)
+
```
nit empty line

oritwas commented on qa/suites/rgw/multisite/all/multi_cluster.yaml:
```diff
@@ -0,0 +1,84 @@
+      client.0:
+        rgw region: jerry
+        rgw zone: seinfeld
+        rgw region root pool: .rgw.region.0
```
We should use `rgw zonegroup root pool ` and not `rgw region root pool`
Also what about` rgw period root pool` and `rgw realm root pool`?

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -349,6 +341,7 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
 
         ctx.daemons.add_daemon(
             remote, 'rgw', client,
+            cluster=cluster_name,
```
make sure it works if cluster_name is empty

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -870,6 +781,7 @@ def configure_regions_and_zones(ctx, config, regions, role_endpoints, realm):
     log.debug('regions are %r', regions)
     log.debug('role_endpoints = %r', role_endpoints)
     log.debug('realm is %r', realm)
+
```
nit

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -893,14 +805,15 @@ def configure_regions_and_zones(ctx, config, regions, role_endpoints, realm):
 
     # clear out the old defaults
-    first_mon = teuthology.get_first_mon(ctx, config)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comments

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -1029,26 +943,35 @@ def pull_configuration(ctx, config, regions, role_endpoints, realm, master_clien
+            log.debug("zone info is: %r", zone_json)
             rgwadmin(ctx, client,
-                     cmd=['zone', 'set', '--rgw-zonegroup', zonegroup,
+                     cmd=['zone', 'set', '--default',
```
why did you change this. this will set and add the zone to the zonegroup

alimaredia replied:
great questions. @cbodley do you remember? I remember there was an issue and this fix solved it, but I can't remember more than that.

cbodley replied:
@oritwas it doesn't look like 'zone set' calls `RGWZoneGroup::add_zone()`, so we had to add the `zonegroup add` below

oritwas replied:
I see ...
I will fix that ...

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -1029,26 +943,35 @@ def pull_configuration(ctx, config, regions, role_endpoints, realm, master_clien
                      check_status=True)
 
             rgwadmin(ctx, client,
+                     cmd=['zonegroup', 'add', '--rgw-zonegroup', zonegroup, '--rgw-zone', zone],
```
no need if you don't change the set command

alimaredia replied:
@cbodley I'm pretty sure this change had to do with the last one for reference.

oritwas commented on qa/tasks/rgw.py:
```diff
@@ -1256,6 +1179,7 @@ def task(ctx, config):
         lambda: create_nonregion_pools(
             ctx=ctx, config=config, regions=regions),
         ]
+    log.debug('Nonregion pools created')
```
missing tab

alimaredia replied:
I think it's fine because it's aligned with subtasks above it and multisite right below it?

oritwas commented on qa/tasks/s3tests.py:
```diff
@@ -29,7 +29,8 @@ def extract_sync_client_data(ctx, client_name):
     return_region_name = None
     return_dict = None
-    client = ctx.ceph['ceph'].conf.get(client_name, None)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
what if cluster_name is empty?

oritwas commented on qa/tasks/s3tests.py:
```diff
@@ -199,19 +200,22 @@ def create_users(ctx, config):
         for section, user in users.iteritems():
             _config_user(s3tests_conf, section, '{user}.{client}'.format(user=user, client=client))
             log.debug('Creating user {user} on {host}'.format(user=s3tests_conf[section]['user_id'], host=client))
+            cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
missing checks

alimaredia replied:
what's the check I'm missing? Did you mean to ask what if the cluster_name is empty?

oritwas replied:
yes , but you explained it is defaults to ceph so that is ok

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -20,14 +19,17 @@ def rgwadmin(ctx, client, cmd, stdin=StringIO(), check_status=False,
              format='json'):
     log.info('rgwadmin: {client} : {cmd}'.format(client=client,cmd=cmd))
     testdir = teuthology.get_testdir(ctx)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -70,8 +72,10 @@ def get_user_successful_ops(out, user):
     return get_user_summary(out, user)['total']['successful_ops']
 
 def get_zone_host_and_port(ctx, client, zone):
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -85,8 +89,10 @@ def get_zone_host_and_port(ctx, client, zone):
     assert False, 'no endpoint for zone {zone} found'.format(zone=zone)
 
 def get_master_zone(ctx, client):
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

oritwas commented on qa/tasks/util/rgw.py:
```diff
@@ -181,3 +189,116 @@ def get_sync_agent(ctx, source):
+    :param client_config: dictionary of client configuration information
+    :returns: zone extracted from client and client_config information
+    """
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
see previous comment

alimaredia commented:
@oritwas @cbodley I removed the multisite directory from qa/suites/rgw. I want to get these changes merged now because they don't, and shouldn't hinder any functionality.

I cleaned up the PR per Orit's last run through the code, but one last runthrough from both of you would be nice. Below are all the runs of teuthology to show that these changes don't break everything:

rgw/verify:
http://pulpito.ceph.com/amaredia-2017-04-05_14:46:26-rgw:verify-wip-ali-testing---basic-smithi/

rgw/multifs (one run for each frontend):
http://pulpito.ceph.com/amaredia-2017-04-03_17:47:50-rgw:multifs-master---basic-mira/
http://pulpito.ceph.com/amaredia-2017-04-04_04:11:13-rgw:multifs-master---basic-vps/

rgw/singleton:
http://pulpito.ceph.com/amaredia-2017-04-04_19:11:15-rgw:singleton-wip-cbodley-testing---basic-mira/

cbodley commented:
@alimaredia for those teuthology runs you linked, were you using the `cluster_name.` syntax in any of the yaml?

alimaredia commented:
@cbodley nope, whatever yaml is existing in the rgw suite now was used

alimaredia commented:
the branch has been rebased and updated, and these thrash tests were run: http://pulpito.ceph.com/amaredia-2017-04-13_18:45:09-rgw:thrash-wip-ali-testing---basic-smithi/

cbodley commented:
thanks @alimaredia, looks good to me

@oritwas can you take a look?
